cirq-core/cirq/experiments/single_qubit_readout_calibration.py

```diff
@@ -96,19 +95,136 @@ def estimate_single_qubit_readout_errors(
         data was finished being collected from the sampler.
     """
+    num_qubits = len(list(qubits))
+    return estimate_parallel_single_qubit_readout_errors(
+        sampler=sampler,
+        qubits=qubits,
+        repetitions=repetitions,
+        trials=2,
+        bit_strings=np.array([[0] * num_qubits, [1] * num_qubits]),
+    )
+
+
+def estimate_parallel_single_qubit_readout_errors(
+    sampler: 'cirq.Sampler',
+    *,
+    qubits: Iterable['cirq.Qid'],
+    trials: int = 20,
+    repetitions: int = 1000,
+    trials_per_batch: Optional[int] = None,
+    bit_strings: np.ndarray = None,
+) -> SingleQubitReadoutCalibrationResult:
+    """Estimate single qubit readout error using parallel operations.
+
+    For each trial, prepare and then measure a random computational basis
+    bitstring on qubits using gates in parallel.
+    Returns a SingleQubitReadoutCalibrationResult which can be used to
+    compute readout errors for each qubit.
+
+    Args:
+        sampler: The `cirq.Sampler` used to run the circuits.
+        qubits: The qubits being tested.
+        repetitions: The number of measurement repetitions to perform for
+            each trial.
+        trials: The number of bitstrings to prepare.
+        trials_per_batch:  If provided, split the experiment into batches
+            with this number of trials in each batch.
+        bit_strings: Optional numpy array of shape (trials, qubits) where the
+            first dimension is the number of the trial and the second
+            dimension is the qubit (ordered by the qubit order from
+            the qubits parameter).  Each value should be a 0 or 1 which
+            specifies which state the qubit should be prepared into during
+            that trial.  If not provided, the function will generate random
+            bit strings for you.
+
+    Returns:
+        A SingleQubitReadoutCalibrationResult storing the readout error
+        probabilities as well as the number of repetitions used to estimate
+        the probabilities. Also stores a timestamp indicating the time when
+        data was finished being collected from the sampler.  Note that,
+        if there did not exist a trial where a given qubit was set to |0〉,
+        the zero-state error will be set to `nan` (not a number).  Likewise
+        for qubits with no |1〉trial and one-state error.
+    """
     qubits = list(qubits)
 
-    zeros_circuit = circuits.Circuit(ops.measure_each(*qubits, key_func=repr))
-    ones_circuit = circuits.Circuit(
-        ops.X.on_each(*qubits), ops.measure_each(*qubits, key_func=repr)
+    if trials <= 0:
+        raise ValueError("Must provide non-zero trials for readout calibration.")
+    if repetitions <= 0:
+        raise ValueError("Must provide non-zero repetition for readout calibration.")
+    if bit_strings is None:
+        bit_strings = np.random.randint(0, 2, size=(trials, len(qubits)))
+    else:
```

Comment:
I think it is worth keeping in, since it avoids a scan of the bit_string array when it is not passed in.  This probably has a slight performance gain.  We don't need to do the sanity checks if we generate the bit_string array.